Refactor IIRFilterNode

b-ma opened the pull request:
Some work on the IIRFilter, mainly:

- simplified the code
- removed the remaining test against the snapshot
- added tests against the `BiquadFilter`,  both for output and get_frequency_response() - there is just one small discrepancy for get_frequency_response and the phases of the notch filter, I didn't really dig why but doesn't seem like a huge issue neither
- handle tail time
 
I guess, if we don't consider the ConvolverNode, that we get near to have #34 and #147 finished now!

b-ma commented:
Huhu, yup that's my "before holiday's sprint" :) It will be more quiet after this week

> I will have a good look soon

No problem take your time, I know I'm just flooding these days

orottier commented on src/node/iir_filter.rs:
```diff
@@ -116,96 +198,66 @@ impl IIRFilterNode {
-        for f in frequency_hz.iter_mut() {
-            *f = f.clamp(min, max);
+            let (mag, phase) = response.to_polar();
+            mag_response[i] = mag as f32;
```
For performance (not really important here) you could zip them along in the loop, this prevents bounds checks on every loop item:
```
for (i, &f, m, p) in frequency_hz.iter().enumerate().zip(mag_response.iter_mut()).zip(phase_response.iter_mut())
```

orottier commented on src/node/iir_filter.rs:
```diff
@@ -278,268 +306,398 @@ impl AudioProcessor for IirFilterRenderer {
-}
+        // eventually resize state according to input number of channels
+        // if in tail time we should continue with previous number of channels
+        if !input.is_silent() {
```
Nitpick, maybe swap the if/else as
```
if input.is_silent {
 // silent case
} else {
 // non-silent
}
```

Parsing the else was a bit hard to me, double negation

orottier reviewed (approved):
Great stuff, nice tests.
As you can see I'm trying to get a benchmark versus main running, will follow soon.
I left some very small nitpicks

orottier commented:
Unfortunately, our own benchmark suite shows the same regression in performance as the github-bot posted.
I have not dug into it, but two things come to mind:
- the `all_channels_equal` trick
- `let states = vec![Vec::<f64>::with_capacity(MAX_CHANNELS); coeffs_len];` is probably better for cache performance when we transpose it `Vec<Vec<coefs>, channels>`, and even better, make it a single Vec, not Vec of Vecs

b-ma commented:
Hey, before you spend too much time on it, I think there is no regression (I really didn't change much the actual computing code) except there is now `output.set_number_of_channels(num_channels);` while it was missing before so it was only mono. If you force `set_channel_number` to 1 in the new version, the footprint should be mostly the same I think

I will try to have a look at your reviews tomorrow

b-ma commented:
Hum actually, I just made a rapid test and the channel count does not explain everything:

### main

```
IIR filter                                                               | 79            | 1519.0x               | 120
```

### refactor

> with default channel_count (stereo)

```
IIR filter                                                               | 160           | 750.0x                | 120
```

> with set_channel_count(1)

```
IIR filter                                                               | 106           | 1132.1x               | 120
```

So, I added or changed something to is not good indeed...

> the all_channels_equal trick

hum, we didn't merge that yet, I made some tests but it wasn't very satisfying (cf. https://github.com/orottier/web-audio-api-rs/issues/117#issuecomment-1206302154)

> let states = vec![Vec::<f64>::with_capacity(MAX_CHANNELS); coeffs_len]; is probably better for cache performance when we transpose it Vec<Vec<coefs>, channels>, and even better, make it a single Vec, not Vec of Vecs

yup probably but that's something I didn't really change neither

I will try to dig into that, sounds weird

b-ma commented:
Actually there seem to be two things that seems to explain (the test are noisy that's a bit difficult to be affirmative):

- the check for subnormal seems to have a cost, but I'm not sure it's a good idea to remove it (or we remove it for now and put that as a task for #194)

- before, there was:
```
let output = self.norm_coeffs[0].0.mul_add(input, self.states[0][channel_number]);
``` 
that I changed (just for readability) to:
```
let b0 = self.norm_coeffs[0].0;
let last_state = self.states[0][channel_number];

let output = b0.mul_add(input, last_state);
```
I'm a bit surprised here, but maybe the conjunction of these two things allowed for some sort of compiler optimizations that we loose, I don't know

On the states layout, before it was `states: Vec<[f64; MAX_CHANNELS]>` that I changed to `states: Vec<Vec<f64>>` which may also be a bad idea. But as you said, we can probably do better than each of these ones.

orottier commented:
Thanks for digging in. I am inclined to remove the subnormal check and leave that for later in the dedicated issue.
I will make some changes and merge if performance is going in the right direction

b-ma commented:
> Thanks for digging in. I am inclined to remove the subnormal check and leave that for later in the dedicated issue.
> I will make some changes and merge if performance is going in the right direction

ok sound goods, for the benches keep in mind that we comparison is biased with main as in main the processing is only mono while it is stereo here (you need to set `iir.set_channel_count(1)` to compare things more properly)

orottier commented:
Yup, when we force mono processing for me there is no performance difference anymore with main..
Going to merge this then, but will roll back the `help compiler optimize inner loop` commit. It makes no difference locally and at CI
